filters/openpolicyagent/openpolicyagent.go

```diff
@@ -0,0 +1,519 @@
+package openpolicyagent
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"text/template"
+	"time"
+
+	ext_authz_v3_core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/config"
+	"github.com/open-policy-agent/opa/logging"
+	"github.com/open-policy-agent/opa/plugins"
+	"github.com/open-policy-agent/opa/plugins/discovery"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/runtime"
+	"github.com/open-policy-agent/opa/storage"
+	"github.com/open-policy-agent/opa/storage/inmem"
+	iCache "github.com/open-policy-agent/opa/topdown/cache"
+	opautil "github.com/open-policy-agent/opa/util"
+	"github.com/opentracing/opentracing-go"
+	"google.golang.org/protobuf/encoding/protojson"
+
+	"github.com/zalando/skipper/filters"
+	"github.com/zalando/skipper/filters/openpolicyagent/internal/envoy"
+	"github.com/zalando/skipper/filters/openpolicyagent/internal/util"
+	"github.com/zalando/skipper/tracing"
+)
+
+const (
+	OpenPolicyAgentDecisionKey        = "open-policy-agent:decision"
+	OpenPolicyAgentDecisionBodyKey    = "open-policy-agent:decision-body"
+	OpenPolicyAgentDecisionHeadersKey = "open-policy-agent:decision-headers"
+)
+
+type OpenPolicyAgentRegistry struct {
+	// Ideally share one Bundle storage across many OPA "instances" using this registry.
+	// This allows to save memory on bundles that are shared
+	// between different policies (i.e. global team memberships)
+	// This not possible due to some limitations in OPA
+	// See https://github.com/open-policy-agent/opa/issues/5707
+
+	instances     map[string]*OpenPolicyAgentInstance
+	refcounts     map[*OpenPolicyAgentInstance]int
+	lastused      map[*OpenPolicyAgentInstance]time.Time
+	mtx           sync.Mutex
```

Comment:
The convention is to put mutex before fields it guards. Maybe also rename it to `mu` like we commonly have it elsewhere.